### ISSUE TITLE
[D8][UX] Import language files at install

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#6660]
## Description of the need

Language (`*.po` files) need to be manually downloaded from the localization server, and manually inserted into a folder in the files directory. 

## Proposed solution

Download language files during install

## Alternatives that have been considered

Continue to install manually

Is there a Drupal or Backdrop contributed module that accomplishes this?
No

## Additional information

Drupal does this.

## Draft of feature description for Press Release (1 paragraph at most)

Backdrop now includes the ability to automatically download language files during the install process, rather than requiring manual download.

Advocating for this issue for the 1.29.0 release: @klonos

**Comment by indigoxela** [comment]
I'm supporting this!

It would be nice to just fetch the file - only in interactive mode, of course.
And even if there's no such translation available, it could be a nice feature to just set the site up in a different language (with language and locale module already prepared for translations...).

BTW, I can see there's a bunch of new issues re language and install and I find that quite confusing. But I guess, that's how things usually go...

The PR isn't really ready yet, as it's obviously just copy-pasted code, incompatible with php 5 and broken, anyway, as it uses function drupal_requirements_severity().
And it's **a lot** of code to accomplish what we'd need. :wink: I don't fully understand, what the PR does, but suspect, we might not need such a big change.

**Comment by docwilmot** [comment]
I did mention, in the issue that inspired this one, https://github.com/backdrop/backdrop-issues/issues/6659#issuecomment-2241450726, that this was early work. I should have said so here. It is in fact mostly cut and paste

**Comment by docwilmot** [comment]
> And it's a lot of code to accomplish what we'd need.

Continuing from my previous comment, FWIW its not excess code at all, I'm pretty sure it is minimal to accomplish what we need. Perhaps we could reduce the amount of error checking in `install_check_translations()`, but not much else. Perhaps you might change your opinion when you actually look at the code.

I do (and did) acknowledge that it is rough (I tend to push early to see what the sandbox complains about and to get the UI out for criticism) and I will be cleaning it up shortly.

**Comment by docwilmot** [comment]
Ready for review

**Comment by indigoxela** [comment]
> Perhaps you might change your opinion when you actually look at the code.

That might be. :grinning:

> (I tend to push early to see what the sandbox complains about and to get the UI out for criticism)

That's what I assumed, anyway. But it is/was really rough. :wink:

**Comment by indigoxela** [comment]
@docwilmot nice cleanup so far! :+1: There's still some phpcs nagging re long lines with arrays and things like that.

I saw, you're using short array syntax, which is OK, as 5.6 is the oldest version Backdrop now even supports. But was that switch to short array syntax by intention (and can we do that now), or was it just copy-paste?

Re overall logic, I'll actually have to test things - which I'll try to do soon.

A more general questions, as that's not really clear to me, yet: is that check-lang-and-download thing supposed to only work in interactive mode, or is it also supposed to work with commandline install?

**Comment by docwilmot** [comment]
https://github.com/backdrop-ops/localize.backdropcms.org/issues/51

**Comment by docwilmot** [comment]
> A more general questions, as that's not really clear to me, yet: is that check-lang-and-download thing supposed to only work in interactive mode, or is it also supposed to work with commandline install?

I have not tested command line installs. I've never done such an install, and I dont think I have the tools to do so. I'll have to hope someone can test that for us.

**Comment by klonos** [comment]
If by command-line installs we mean `bee si`, then I have scripts that are doing that all the time for me on my local - just never tried to specify language and always do it in English. If that is all it needs with regards to testing, then I can do that.

**Comment by indigoxela** [comment]
So cool, l10n_server now has such end point to directly query available languages! :tada:

**Comment by klonos** [comment]
I've done a first-pass review of the code and added comments/suggestions/questions for things that caught my eye. I'll do some actual testing on my local next.

**Comment by klonos** [comment]
I've tested with a couple of non-English languages, trying both RTL and LTR, and everything seems to be working as expected 👍🏼 ...only a couple of minor things that could be improved:
- Since I am on holidays and on a slow internet connection, the time it takes between clicking "Save and continue" and the actual installation to start was noticeable (10-20 sec). Would be nice to have some "downloading language file" indication going on while that is happening.
- When selecting a RTL language, the direction of the text does not switch to RTL until the "Site configuration" step. Would be nice if the switch happened instantly, either when selecting the language from the select menu, or at least when clicking "Save and continue".

Otherwise, this looks great! 👏🏼 ❤️

**Comment by stpaultim** [comment]
Just testing this PR and realizing how bad the current situation is. As has been pointed out, by default we ask which language a person wants to use when spinning up a site, but then only provide one option (along with instructions on how to add other languages). 

;-(

![image](https://github.com/user-attachments/assets/fff4eb4c-5e75-41cd-9df8-b7e3a2eb6786)

I tested the PR and spun up sites in German and Spanish. It worked so smoothly and just makes sense!

Big thanks to @docwilmot for this PR.

This is the same page after PR. No extra steps to spin up a site in another language!

![image](https://github.com/user-attachments/assets/7a4f5451-074b-4b6c-aa6b-3cf5282748a8)

**Comment by klonos** [comment]
I just gave the PR another spin after the commits with regards to pulling the list of languages directly from localize.b.org have been merged, and confirm that the main functionality is still working as expected 👍🏼  ...I haven't gone to the extend of simulating things like the localization server being offline or errors during the downloading of .po files or directories missing or not being writable etc., but the logic around these things LGTM.

I've found some minor nits in the PR (wrong indentation, extra dots etc.), and there's the issue of the shorthand array format being used, which @indigoxela also mentioned in an earlier review.

The two previous points I've made still stand:
- No indication that a language file is being downloaded (issue more obvious in slow internet connections). Should be able to reproduce by simulating network throttling via browser dev tools.
- RTL languages don't switch the installer screens to RTL until it reaches the "Site configuration" form - should be straight after selecting a RTL language and clicking "Save and continue". The json output at ??? already provides `"direction": "1"` for such languages, so it should be doable, but happy to leave it for a follow-up issue.

If nobody else wants to be the official advocate for this feature request, then I'm happy to do it and set the milestone.

**Comment by indigoxela** [comment]
Finally, some time for testing this. :wink: 

I can't express, how much I like this! :heart: It's how it should have been and (hopefully) will be as of September.
I'll do some more testing, just wanted to quickly share my enthusiasm.

@docwilmot there's still phpcs nagging (6). Not sure - don't you see these comments when visiting https://github.com/backdrop/backdrop/pull/4834/files
And also a misleading return type (easy to fix).

**Comment by docwilmot** [comment]
Thanks for testing. I don't quite know how to address PHPCS nagging about `st()` strings though? We're meant to keep them without line breaks, no?

**Comment by indigoxela** [comment]
> We're meant to keep them without line breaks, no?

We don't break inside the function, where it would cause trouble, but we break inside the array. I added a simplified example to one of the nagging comments.

**Comment by indigoxela** [comment]
Now some nasty testing: I closed outgoing traffic to localize server via firewall.

Hmmm not yet ideal.
First a  long waiting time (connect timeout could be smaller, I guess), but the actual problem:

![fail-on-no-access](https://github.com/user-attachments/assets/5579bf5f-7e0c-4fe5-8dc5-1f7aef32b596)

We should catch that instead of TypeError. :wink: 
```
TypeError: count(): Argument #1 ($value) must be of type Countable|array, bool given in install_select_language() (line 1344 of /var/www/bdroptesting/html/core/includes/install.core.inc). 
```
And several other comments re fallback behavior when no connection to server.

I don't think, the lack of internet connection should be a fatal error for installation. Just drop a friendly message and move on with installation. :wink:

**Comment by argiepiano** [comment]
> I don't think, the lack of internet connection should be a fatal error for installation

True! I have at times installed new Backdrop development sites in places like airplanes with no wifi.

**Comment by docwilmot** [comment]
Responded on the PR. There was a stray `else` clause in there that was allowing a `FALSE` to be returned, if no internet. I'm sure I tested that. There should be no error at all without network. If no netwrok, the code checks for downloaded files, if no files, assumes English.

**Comment by indigoxela** [comment]
> Responded on the PR. There was a stray else clause...

Ah got it. :+1: 

> I'm sure I tested that.

So am I :wink:. Such oversights are there to keep us testers and reviewer vigilant, right? :grinning:
Anyway, take your time to update the PR, maybe I'll do more testing tomorrow. :wave: No need to rush.

**Comment by klonos** [comment]
@docwilmot I'm assuming that you are advocating for this feature for the next minor release. Right? Can you please either add a note to the issue summary, indicating that you are indeed advocating for this, or let us know if you want to advocate for something else? There's plenty of folks excited about this, so I'm sure that one of us will be glad to advocate instead. With the feature freeze coming up in about 4 weeks, I'd like to set the milestone, so that this is kept within our radar 😉

**Comment by docwilmot** [comment]
My attendance is unreliable so please advocate on my behalf.

**Comment by indigoxela** [comment]
@klonos if an advocate is necessary, I could do that as well. Or you could. Or we both could be advocates together. :wink: 
If it's about the milestone: yes, 1.29 is the goal. :muscle:

**Comment by klonos** [comment]
@indigoxela since we only allow advocating for one feature at a time, and since I'm assuming you'll be advocating for #6588, I'll advocate for this one here 😉 ...added myself as advocate here, so also adding the milestone.

**Comment by indigoxela** [comment]
> we only allow advocating for one feature at a time

Oops, didn't know that. However, I'm not currently advocating for any issue.

> I'm assuming you'll be advocating for https://github.com/backdrop/backdrop-issues/issues/6588

I'm *providing the PR* for that issue, so I'm not advocating. I think, that's different roles. Actually I'm still not sure if the overall interest for that feature (other PR) is enough to get it into 1.29. :wink: But that's off topic here.

The language automatic fetch feature rocks, and I'm sure it's a feature lots of people would love to see going in. :wink: Including me.

**Comment by indigoxela** [comment]
Some more testing today:

**Test 1:** with full internet connection
Works like a charm. Monolingual in my language. :+1: 

**Test 2:** without connection to backdropcms server from beginning
There's a slight delay, but everything works smoothly, English install. :+1: 
Unrelated: the check for the update server (not new in this PR) seems to wait for the full default timeout... That's something for a follow-up, though (for update/update.fetch.inc).

**Test 3:** select language and then close internet connection to translation server before next step.
Well, that's... not how it should be, although this is an edge case.
```
Choose language

SQLSTATE[42S02]: Base table or view not found: 1146 Table 'bdroptesting.tempstore' doesn't exist
```
Not sure, how we'd ever get to the prepared messages, that would inform us about failed translation download.
Instead of `if (!$online) {...` and then showing a message, the install seems to fail hard.

**Comment by indigoxela** [comment]
Yet more testing, now commandline:

**Test 4:** Commandline install with language set as option, internet off
```
An error occurred. Output of installation attempt is as follows:
Internet: The translation server is offline or cannot be reached.
...
```
Yes, that's expected and works flawlessly. No install attempted with proper message. Only the "translations" directory got created. :+1: 

Test 5: Commandline install with language set as option, _with_ internet
Flawless install in correct language. :+1:

**Comment by indigoxela** [comment]
@docwilmot all of my current comments on your PR can get closed as resolved - I can't because of Github perms, but you can.

The only remaining "problem" is the weird edge case with intermittent network failure. We could dig deeper into that, or we could ignore it - as the edge case it is. What do you think?

**Comment by klonos** [comment]
Thank you for the very thorough testing @indigoxela!! 🙏🏼 🙏🏼 🙏🏼 

I have left one suggestion and one question in the PR, but I believe that that's it from me.

I would also like us to address the issue with the wait time after we click the "Save and continue" button in slow connections (I suggested adding some progress bar or other indication that something is actually happening in the background), but this can also be a follow-up, like the edge case @indigoxela mentioned. This is looking really good! @docwilmot! 👍🏼 ❤️

**Comment by docwilmot** [comment]
> Test 3: select language and then close internet connection to translation server before next step.

Tested we could prevent that error by using `backdrop_build_form()` instead of `backdrop_get_form()` and setting `$form_state['no_cache']` but then the form just reloads with a proper form error that the language key doesnt exist in the select list so not much help, but less ugly error.

**Comment by indigoxela** [comment]
> ... we could prevent that error by using backdrop_build_form() ... but then the form just reloads with a proper form error that the language key doesnt exist in the select list so not much help, but less ugly error.

@docwilmot  Not sure. As you state - not much help, either. :shrug:

**Comment by indigoxela** [comment]
A final round of testing from my side: works like a charm. :+1: 

With firewall closed, there's still a slight delay caused by the update server check (unrelated to language), which is also logged to dblog.

Works for me, and the code also looks good, but I'll leave the RTBC label to @klonos, who also did a lot of review here.

**Comment by argiepiano** [comment]
Love this and also works for me 👍🏽 . Thanks for all the work here

**Comment by herbdool** [comment]
I've reviewed the code. Just a couple minor changes, otherwise looks good.

**Comment by docwilmot** [comment]
Updated from @herbdool review, should be good for final review now.

**Comment by herbdool** [comment]
LGTM!

**Comment by laryn** [comment]
This is great and a wonderful way to roll out the welcome mat for folks whose first language may not be English. I tested and looked over the code but it looks like the heavy lifting was all done. 

Thanks @docwilmot for filing the issue and putting in the work on the PR, @indigoxela and @klonos for the thorough, consistent feedback and testing, @herbdool for code review, @stpaultim and @argiepiano for your testing and inputs! Merged into `1.x` to be included in `1.29.x`, hopefully with time yet for any follow-ups that may be desired.
